### DISCUSSION TITLE
Can a stream which is not "Started" call onAudioReady?

### DISCUSSION THREAD

**Discussion by ashfaq-opn** [google/oboe#1226] [Q&A] [answered]
I have a small question. If a stream is not in the state of "Started", can it call onAudioReady callback?

More specifically, should I check like below,

```if (stream->getState() == oboe::StreamState::Started) {```

Before any IO operation?

**Comment by philburk** [comment, marked as answer]
You will **not** get an onAudioReady callback before you call stream->requestStart().
After requestStart() the stream will be in state Starting state then it will transition to Started state.
A callback can occur any time after requestStart().
It is possible for a callback to still be running after calling requestStop() because that is an asynchronous request.

> Before any IO operation?

What kind of IO operation? You do not need to write the stream. Just fill the buffer in onAudioReady().
Generally you do not need to check the state. That is mostly for debugging.

**Reply by ashfaq-opn** [reply]
By IO operation I meant file IO operation like saving to a file or reading from a file.

I got the answer, I don't need that check before those operations. Thank you very much.

**Comment by philburk** [comment]
Note that the audio callbacks are real-time threads. So you should not be doing any blocking operations. This include:

- Opening, reading, writing or closing files.
- Allocating or freeing memory.
- Any network operation.

You should open a stream without a callback, PerformanceMode::None.
Then read() or write() the audio stream in a loop while reading or writing the file.

Or, if you need a low latency callback, read or write a large FIFO and then have a background thread that writes or reads the FIFO and does the file I/O.

**Reply by ashfaq-opn** [reply]
Thanks for these kind suggestions.

I am using non-blocking FIFO queues to perform the file write operations.

I am reading from data loaded into RAM, so that operation is non-blocking anyway. :)
